DO-1675 - EC2 Instance Retention

tomislacker opened the pull request:
- Adds a script to manage EC2 instance retention
  - Intended for execution via [AWS Lambda](https://aws.amazon.com/lambda/) but can be executed via CLI
  - Respects termination protection and quietly ignores those instances
  - CLI arguments _(requires [docopt](http://docopt.org/))_:
    
    ``` sh
      $ python instancecleanup/instancecleanup.py --help                                     
      instancecleanup.py
    
      Usage:
          instancecleanup.py [options]
    
      Options:
          -h, --help          Show this screen
          --version           Show the script version
          --hot-run           Actually terminate instance(s)
          --days=<D>          Number of days since the instance was stopped
          --log-level LEVEL   Logging level (see logging module)
    ```
- Fixes a small typo

Commit:
Sync up to common .gitignore for Python projects

Commit:
Add Makefile for ease of use

Commit:
Fix typo

Commit:
[DO-1675] Add lambda function to cleanup stale instances

Commit:
[DO-1675] Add IAM policy document for listing & terminating EC2 instances
As of 2016-09-26 this is untested

Commit:
[DO-1675] Add documentation for instancecleanup

Commit:
Fix interpreter for ease of development

tomislacker commented:
@shawnsi @speshak 

We probably don't want to activate this until we've had a chance to manually prune the list down a bit but I'm not sure. Here's a sample of what we'd be getting rid of at different default retention times (since most instances don't have the `ops:retention` tag):

| Retention Days | Instance Count |
| --- | --- |
| 1095 _(3yrs)_ | 3 |
| 730 _(2yrs)_ | 18 |
| 365 _(1yr)_ | 28 |
| 182 _(6mo)_ | 29 |

Commit:
Make upload.py executable

shawnsi commented:
@tomislacker @speshak I'm considering dropping default retention outright.  That or setting it to infinite by default.  What do you think?

tomislacker commented:
@shawnsi I think that's a safer idea in general. I can modify all the functions to have that behavior and then if a retention is specifically passed, it'd treat that as a default retention.

speshak commented:
Big 👍  for dropping default retention.

shawnsi commented:
@tomislacker go with that.

Commit:
Remove default retention, do not implicitly remove any resources

Squashed commit of the following:

commit c64fcd7bd46834379e778ebcb9d66759215743ce
Author: Ben Tomasik <Ben.Tomasik@meredith.com>
Date:   Wed Sep 28 10:54:49 2016 -0500

    [DO-1675] Remove default retention from instancecleanup

commit f33ff985592f7dae474bdd04a40b57ed98f73bb0
Author: Ben Tomasik <Ben.Tomasik@meredith.com>
Date:   Wed Sep 28 10:04:25 2016 -0500

    [DO-1675] Remove default retention for snapshotcleanup

commit 1c0aeea951162a60754d848059f3d8d078677a04
Author: Ben Tomasik <Ben.Tomasik@meredith.com>
Date:   Wed Sep 28 10:00:45 2016 -0500

    [DO-1675] Remove default retention for amicleanup

tomislacker commented:
@speshak @shawnsi the default retention period has been nullified by allowing that value to be set to `None`. This results in only resources with the `ops:retention` tag set explicitly will have the potential for being removed during an invocation.

Commit:
[DO-1675] Quietly ignore termination errors due to termination protection

tomislacker commented:
This has been marked WIP as it should terminate instances that are running as well, not just stopped, all based on the `ops:retention` tag. Then rely on operations to apply termination protection as needed.

Commit:
[DO-1675] Terminate instances that are running if they don't have termination protection

Commit:
[DO-1675] Remove instances based on creation time rather than stopped time
This heavily relies on the usage of terminate protection to ensure important instances do not get cleaned up

tomislacker commented:
@speshak @shawnsi review?

Commit:
[DO-1675] Only consider stale instances w/o termination protection

tomislacker commented:
With the removal of a default retention period, and _only_ relying on the termination of instances with an `ops:retention` tag that do not have termination protection enabled, we've reached the point where the lambda job will not execute any terminations.

shawnsi reviewed (approved):
LGTM

tomislacker commented:
@shawnsi you'll have to execute the merge as I don't have write-access to this repo.
